### ISSUE TITLE
Implement UIDS 4.x accordion changes

### ISSUE THREAD

**Issue by GaryRidgway** [uiowa/uiowa#4004]
## Stories
Implement the changes to the accordion component from UIDS 4

## Proposed solution
- Address uiowa/uids#961
- Update template
- Review any custom CSS related to accordions in SiteNow
- Search for instances of `'@uids_base/uids/accordion-item.html.twig'` to see where this is getting implemented. Review these to see if we need adjust our implementations.

<details>
<summary>Previous information</summary>
Story: As a website visitor, I can interact with accordions that are as accessible as possible.
Proposed solution:
Change the template for accordions to utilize the details tag like below.

https://www.w3schools.com/tags/tag_details.asp
<details>
<summary> Apparently details tags work in markdown EXACTLY as they do in HTML.</summary>
<div>I knew they worked in Github, but not that they were exactly the same! Pretty neat!</div>
</details>
</details>

**Comment by joewhitsitt** [comment]
Based on what was discovered in https://github.com/uiowa/uiowa/pull/5277, I think it would be better to address https://github.com/uiowa/uiowa/issues/5106 as part of this issue, by simplifying the nesting and not negating the a11y of the headings.

**Comment by bspeare** [comment]
I'm not seeing the `<details>` in use on https://www.w3.org/WAI/ARIA/apg/patterns/accordion/.   I also don't see it on https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/.   Do we really want to convert to `<details>` or should we just compare our version with https://www.w3.org/WAI/ARIA/apg/patterns/accordion/?

**Comment by bspeare** [comment]
Looks like they are aware of the issue: https://github.com/w3c/aria-practices/issues/1856

**Comment by joewhitsitt** [comment]
I believe the origin of this was a tip from TM that `<details>` was accessible out of the box and that the show/hide functionality was free of charge. The accordion example on w3 doesn't have all of the tablist role information which is what is causing the flag in https://github.com/uiowa/uiowa/issues/5106

The tablist was also something I picked up from TM a long time ago related to collapsible menus.

If we want to use this issue to revisit what an accordion should be I think that can be in scope, but we should go the extra distance to get agreement from others on campus.

**Comment by bspeare** [comment]
Great points @joewhitsitt, I'm fine with either direction.

**Comment by mcking65** [comment]
Ages ago there was the idea that tablist would be a way to build accordions. We discarded that idea when we built the APG accordion. tablist and accordion are usually incompatible because the element structure and reading order of a tabs pattern and an accordion are fundamentally different.

Tabs structure is:

~~~
<tablist><tab1/><tab2/><tab3/></tablist>
<tabpanel>panel 1 content</tabpanel>
<tabpanel>panel 2 content</tabpanel>
...
~~~

Whereas an accordion structure is:
~~~
<according header 1/>
<panel 1/>
<accordion header 2/>
<panel 2/>
...
~~~

It just doesn't work if you try to group all the accordion header headers together into a tablist. Tabs need to be inside of a tablist, and you can't put panels inbetween the tabs.

**Comment by GaryRidgway** [comment]
[I have made a code example here that uses our styles on an accordion made with the `<details>` tag.](https://codepen.io/purple_salmon/pen/VwXaLaw) I know it is a bit messy, but it is a proof of concept that this will work, and it will use a lot less JS.

**Comment by quamsta** [comment]
+1 to using `<details>` element. One thing that's really nice about them is that you can Ctrl + F or Cmd + F in browser and find contents within them easily (they auto expand in this scenario) in most browsers. Something that's really handy and I know @benlewzz has asked for in the past

**Comment by joewhitsitt** [comment]
```html
<div class="accordion-container" aria-labelledby="accordion-heading">
  <h2 id="accordion-heading" class="accordion-heading">Accordion group</h2>
  
  <details class="accordion" aria-labelledby="system-requirements-heading">
    <summary id="system-requirements-heading" class="accordion__heading">
      <span class="accordion__button">
        System Requirements
      </span>
    </summary>
    <div class="accordion__content">
      <p>Requires a computer running an operating system. The computer
    must have some memory and ideally some kind of long-term storage.
    An input device as well as some form of output device is
    recommended.</p>
    </div>
  </details>
</div>
```

Working off of @GaryRidgway's example in https://github.com/uiowa/uiowa/issues/4004#issuecomment-1178250260, would requiring a grouping heading be the solution here?

**Comment by GaryRidgway** [comment]
## Proposed solution:

1. Update UIDS to accommodate change to `Details` https://github.com/uiowa/uids/pull/943
2. We need to change the template of the accordion items to use the `details` html element.
3. We need to move unique IDs from the `H2` elements and move them to the `summary` element. We can then move the `H2` elements inside the `summary` element to keep the appearance and heading structure.
4. The container element for the collection needs to have the `aria-owns` attribute, referencing the `summary` ids of each element in the collection, like so: `aria-owns="acccordion-1 accordion-2 accordion-3"`
5. Make sure that the collection container has the aria tablist functionality. The collection should be structured so that the tablist owns all the different accordions. *(This may be repetetive of item 1. if so, you may disregard.)*

**Comment by briand44** [comment]
Grooming conversation... A question was brought up around #3 ^^ on whether this would be implemented through JS or through twig. We would like to add IDs to the collection element via PHP preprocessing.

Ben mentioned:
>I’ve got a UIDS pr with accordion migration in it ready for review.  Do we want a slice to do it in UIDS4 first? https://github.com/uiowa/uids/pull/943

Scoped this slice to include both UIDS work to accommodate change to `details` as well as the work in uids_base.

**Comment by GaryRidgway** [comment]
> ```
> <div class="accordion-container" aria-labelledby="accordion-heading">
>   <h2 id="accordion-heading" class="accordion-heading">Accordion group</h2>
>   
>   <details class="accordion" aria-labelledby="system-requirements-heading">
>     <summary id="system-requirements-heading" class="accordion__heading">
>       <span class="accordion__button">
>         System Requirements
>       </span>
>     </summary>
>     <div class="accordion__content">
>       <p>Requires a computer running an operating system. The computer
>     must have some memory and ideally some kind of long-term storage.
>     An input device as well as some form of output device is
>     recommended.</p>
>     </div>
>   </details>
> </div>
> ```
> 
> Working off of @GaryRidgway's example in [#4004 (comment)](https://github.com/uiowa/uiowa/issues/4004#issuecomment-1178250260), would requiring a grouping heading be the solution here?

I think it should be. I think it would make the multiselectable stuff work better too.

**Comment by joewhitsitt** [comment]
@GaryRidgway I should note that this comment was made before our meeting with TM as a details example without the tablist stuff. We have since decided to not require the collection headline and strive for the aria-owns information.

**Comment by GaryRidgway** [comment]
> @GaryRidgway I should note that this comment was made before our meeting with TM as a details example without the tablist stuff. We have since decided to not require the collection headline and strive for the aria-owns information.

Ah, understood, thank you for clarifying!

**Comment by bspeare** [comment]
I've removed Alan's assignment from this and moved this back to the planned column. The UIDS work from this at https://github.com/uiowa/uids/pull/955 is now ready to be integrated into uids_base for the second part of this planned work.

**Comment by pyrello** [comment]
Taking over this work as @GaryRidgway is now on vacation.

**Comment by pyrello** [comment]
This has been implemented in UIDS and we are ready to do the work in SiteNow.

**Comment by pyrello** [comment]
We should address uiowa/uids#961 as part of these updates.

**Comment by briand44** [comment]
Grooming conversation... New UIDS PR to address fixes in https://github.com/uiowa/uids/issues/961. Apply structural changes from UIDS into uids_base. There are a few different places we'll want to look at in uids_base. We'll want to take a look at layout builder styles if there are any used by accordions. HR was brought up as work we may want to address the FAQ work first. If we prioritize this before HR work, let's not worry about HR as part of this.

**Comment by bspeare** [comment]
Since https://github.com/uiowa/uids/pull/955 was merged, we are blocked from merging in the banner work https://github.com/uiowa/uiowa/pull/8343 until this accordion integration is completed.

**Comment by joewhitsitt** [comment]
Mmmkay. uids changes are in. I've been trying to wrangle things in uids_base and will push something up once it doesn't look completely whack.

**Comment by joewhitsitt** [comment]
It looks like the UIDS styles only target h2s and not any header (e.g. h3, h4, etc.). Based on the setting in the collections block. Should I modify to account for this in uids_base or make a change to uids? `:is(h2, h3, h4, h5, h6) {`?

**Comment by bspeare** [comment]
> It looks like the UIDS styles only target h2s and not any header (e.g. h3, h4, etc.). Based on the setting in the collections block. Should I modify to account for this in uids_base or make a change to uids? `:is(h2, h3, h4, h5, h6) {`?

You should make them in UIDS.

**Comment by bspeare** [comment]
You can update the hash at https://github.com/uiowa/uiowa/blob/main/docroot/themes/custom/uids_base/package.json#L6 and then run `ddev yarn upgrade @uiowa/uids` and then `blt frontend`.

**Comment by joewhitsitt** [comment]
<details>
More damned if you do, damned if you don't. [Just implemented aria-owns on transfer-tips-list](https://github.com/uiowa/uiowa/pull/8405/commits/73b083f7d503f2f762d63395b72f41ec3ed97bc0) and now a nested list inside of the accordion item with the accordion having aria-owns information cause violations in AInspector..

<img width="1596" alt="Screenshot 2025-01-14 at 4 51 37 PM" src="https://github.com/user-attachments/assets/613e424b-a792-41a1-8bc5-3e0ef9e15d4e" />
</details>

UPDATE: Nevermind. I had the wrong ids in aria-owns. It is meant for the `<summary>` elements. https://github.com/uiowa/uiowa/pull/8405/commits/3d04bfc1ef0ae297fb914d3ea20bb059ad0b6c97

**Comment by joewhitsitt** [comment]
Ran into a new issue where uipress book type field template (which has the parent accordion stuff) is being used for related reads section but the paragraphs items inside are using a different viewmode so it is missing the related tablist markup. Need to figure out how to get a generic field template for this running or switch the node field render to use a custom template.

<img width="725" alt="Image" src="https://github.com/user-attachments/assets/b10b6019-20b1-4577-84e2-fe35bb6f5f2a" />

**Comment by joewhitsitt** [comment]
I got a little distracted during the wcag meeting thinking about what @pyrello and I talked about earlier in the day yesterday. Started to think about how we could include something like this in order to defend our choices around accessibility and in order to make sure we don't introduce regressions as we implement or improve the component. Probably overkill, but I wanted to start to put something together to get feedback.

<details>

# Accordion Accessibility Notes

Last updated: January 2025
Compliance Level: WCAG 2.1 AA
Component: [Accordion](https://uids.brand.uiowa.edu/latest/?path=/docs/components-accordion--docs)


## Rationale

Purpose: To provide an accessible Accordion component that supports a wide range of users, including those using assistive technologies.

Approach: Combine ARIA roles, properties, and HTML5 elements to ensure a robust and inclusive user experience.

## Key Accessibility Features

Proper heading structure: Through our main implementation (Collection), control of the heading hierarchy/nesting is managed by the editor.

Grouping: The Accordion items are grouped together using `role="tablist"` and information about the group is provided using `aria-owns`.

Identification: Each Accordion item is marked with `role="tab"` and is associated with its content using the `aria-owns` attribute on the parent `tablist` container.

Tab Navigation: Users can navigate through Accordion items using the Tab key.

Activation: Accordion items can be expanded or collapsed using keystrokes specific to the browser or tool used to consume the webpage.

Visibility: Content within the Accordion items is not hidden from in-page browser search functionality, ensuring that users can find content using browser search features.

Unique IDs: Through our main implementation (Collection), items can be supplied with custom IDs for anchor link use.

## Accessibility Testing

This combination of strategies was developed with consultation from ITS' ISPO's T.M. Weissenberger.

Developer screen reader testing was performed using Apple's VoiceOver tool.

Browser accessibility extensions were used to scan for compliance issues against WCAG 2.1 AA, including:

- AInspector
- Google's Lighthouse
- WebAIM's WAVE
- SiteImprove

## Browser Compatibility

Tested across modern browsers (e.g., Chrome, Firefox, Safari) for consistent behavior.

## Known Issues

Issue 1: Uses ARIA roles on incompatible elements

Many HTML elements can only be assigned certain ARIA roles. Using ARIA roles where they are not allowed can interfere with the accessibility of the web page.

### Examples:

- Use of `role="none"` on `<details>`
- Use of `role="tab"` on `<summary>`

</details>
